src/integration-test/java/com/commercetools/sync/integration/commons/utils/ITUtils.java

```diff
@@ -42,44 +39,50 @@ public static void deleteTypesFromTargetAndSource() {
-        throws JsonProcessingException {
-        final ObjectMapper mapper = new ObjectMapper();
-        return mapper.writeValueAsString(statistics);
+    public static <T extends Resource, C extends QueryDsl<T, C>> void queryAndExecute(
+        @Nonnull final SphereClient ctpClient,
+        @Nonnull final Supplier<QueryDsl<T, C>> queryRequestSupplier,
+        @Nonnull final Function<T, SphereRequest<T>> resourceToRequestMapper) {
+
+        queryAndCompose(ctpClient, queryRequestSupplier,
+            resource -> ctpClient.execute(resourceToRequestMapper.apply(resource)));
     }
 
     /**
-     * Applies the {@code pageMapper} function on each page fetched from the supplied {@code queryRequestSupplier} on
-     * the supplied {@code ctpClient}.
+     * Applies the {@code resourceToStageMapper} function on each page, fetched from the supplied
+     * {@code queryRequestSupplier} on the supplied {@code ctpClient}, to map each resource to a {@link CompletionStage}
+     * and then executes these stages in parallel within each page.
      *
-     * @param ctpClient            defines the CTP project to apply the query on.
-     * @param queryRequestSupplier defines a supplier which, when executed, returns the query that should be made on
-     *                             the CTP project.
-     * @param resourceMapper       defines a mapper function that should be applied on each resource in the fetched page
-     *                             from the query on the specified CTP project.
+     * @param ctpClient             defines the CTP project to apply the query on.
+     * @param queryRequestSupplier  defines a supplier which, when executed, returns the query that should be made on
+     *                              the CTP project.
+     * @param resourceToStageMapper defines a mapper function that should be applied on each resource, in the fetched
+     *                              page from the query on the specified CTP project, to map it to a
+     *                              {@link CompletionStage}.
      */
-    public static <T extends Resource, C extends QueryDsl<T, C>> void queryAndApply(
+    public static <T extends Resource, C extends QueryDsl<T, C>, S> void queryAndCompose(
         @Nonnull final SphereClient ctpClient,
         @Nonnull final Supplier<QueryDsl<T, C>> queryRequestSupplier,
-        @Nonnull final Function<T, SphereRequest<T>> resourceMapper) {
+        @Nonnull final Function<T, CompletionStage<S>> resourceToStageMapper) {
 
-        final Function<List<T>, Stream<CompletableFuture<T>>> pageMapper =
-            pageElements -> pageElements.stream()
-                                        .map(resourceMapper)
-                                        .map(ctpClient::execute)
-                                        .map(CompletionStage::toCompletableFuture);
+        queryAll(ctpClient, queryRequestSupplier.get(), resourceToStageMapper)
```

Comment:
missed to write a comment before approval:
`Supplier<QueryDsl<T, C>> queryRequestSupplier`
supplier here looks redundant since it is called immediately and directly without any conditions.

Same for `queryAndExecute` signature above

Reply:
8e59755